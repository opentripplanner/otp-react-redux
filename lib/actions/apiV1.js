import qs from 'qs'

import {
  createQueryAction,
  findGeometryForPattern,
  findGeometryForTrip,
  findPatternsForRouteError,
  findPatternsForRouteResponse,
  findRouteError,
  findRouteResponse,
  findRoutesError,
  findRoutesResponse,
  findStopsForPattern,
  findStopsForTrip,
  findStopTimesForTrip,
  findTripError,
  findTripResponse,
  receivedVehiclePositions,
  receivedVehiclePositionsError
} from './api'
import { setViewedStop } from './ui'

// Import must be done like this as maplibregl is incompatible with jest
let maplibregl = null
if (typeof jest === 'undefined') maplibregl = require('maplibre-gl')

const findTrip = (params) =>
  createQueryAction(
    `index/trips/${params.tripId}`,
    findTripResponse,
    findTripError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        dispatch(findStopsForTrip({ tripId: params.tripId }))
        dispatch(findStopTimesForTrip({ tripId: params.tripId }))
        dispatch(findGeometryForTrip({ tripId: params.tripId }))
      }
    }
  )

export function vehicleRentalQuery(
  params,
  responseAction,
  errorAction,
  options
) {
  const paramsString = qs.stringify(params)
  const endpoint = `vehicle_rental${paramsString ? `?${paramsString}` : ''}`
  return createQueryAction(endpoint, responseAction, errorAction, options)
}

/**
 * Stop times for stop query (used in stop viewer).
 */
export function findStopTimesForStop(params) {
  return function (dispatch, getState) {
    console.warn('OTP1 is not supported.')
  }
}

const getVehiclePositionsForRoute = (routeId) =>
  createQueryAction(
    `index/routes/${routeId}/vehicles`,
    receivedVehiclePositions,
    receivedVehiclePositionsError,
    {
      noThrottle: true,
      rewritePayload: (payload) => {
        return {
          routeId: routeId,
          vehicles: payload
        }
      }
    }
  )

export const findPatternsForRoute = (params) =>
  createQueryAction(
    `index/routes/${params.routeId}/patterns?includeGeometry=true`,
    findPatternsForRouteResponse,
    findPatternsForRouteError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        // load geometry for each pattern
        payload.forEach((ptn) => {
          // Some OTP instances don't support includeGeometry.
          // We need to manually fetch geometry in these cases.
          if (!ptn.geometry) {
            dispatch(
              findGeometryForPattern({
                patternId: ptn.id,
                routeId: params.routeId
              })
            )
          }
          dispatch(
            findStopsForPattern({
              patternId: ptn.id,
              routeId: params.routeId
            })
          )
        })
      },

      rewritePayload: (payload) => {
        // convert pattern array to ID-mapped object
        const patterns = {}
        payload.forEach((ptn) => {
          patterns[ptn.id] = ptn
        })

        return {
          patterns,
          routeId: params.routeId
        }
      }
    }
  )

export const findRoute = (params) =>
  createQueryAction(
    `index/routes/${params.routeId}`,
    findRouteResponse,
    findRouteError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        // load patterns
        dispatch(findPatternsForRoute({ routeId: params.routeId }))
        dispatch(setViewedStop(null))
      }
    }
  )

export function findRoutes() {
  return createQueryAction(
    'index/routes',
    findRoutesResponse,
    findRoutesError,
    {
      postprocess: (payload, dispatch) => {
        dispatch(setViewedStop(null))
      },
      rewritePayload: (payload) => {
        const routes = {}
        payload.forEach((rte) => {
          routes[rte.id] = rte
        })
        return routes
      },
      serviceId: 'routes'
    }
  )
}

export function routingQuery(searchId = null, updateSearchInReducer) {
  return function (dispatch, getState) {
    console.warn('OTP1 is not supported.')
  }
}
export default {
  findPatternsForRoute,
  findRoute,
  findRoutes,
  findStopTimesForStop,
  findTrip,
  getVehiclePositionsForRoute,
  routingQuery,
  vehicleRentalQuery
}
