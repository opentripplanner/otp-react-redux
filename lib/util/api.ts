import { convertModeSettingValue } from '@opentripplanner/trip-form'
import { ModeSetting, ModeSettingValues } from '@opentripplanner/types'
import { toDate } from 'date-fns-tz'
import coreUtils from '@opentripplanner/core-utils'
import qs from 'qs'

const { getUrlParams } = coreUtils.query
const { getCurrentDate, getCurrentTime } = coreUtils.time

export const SERVICE_BREAK = '03:30'

/**
 * Generates a record of all mode setting keys and their values from (in order of priority):
 * URL Search Params, Initial Value, and default from definition.
 * @param urlSearchParams URL Search Parameters containing mode setting values
 * @param modeSettingDefinitions Complete listing of mode setting definitions with defaults
 * @param initalStateValues Initial state values
 * @returns Record of mode setting keys and their values
 */
export const generateModeSettingValues = (
  urlSearchParams: URLSearchParams,
  modeSettingDefinitions: ModeSetting[],
  initalStateValues: ModeSettingValues
): ModeSettingValues => {
  // Get mode setting values from the url, or initial state config, or default value in definition
  const modeSettingValues = modeSettingDefinitions?.reduce<ModeSettingValues>(
    (acc, setting) => {
      const fromUrl = urlSearchParams.get(setting.key)
      acc[setting.key] = fromUrl
        ? convertModeSettingValue(setting, fromUrl)
        : initalStateValues?.[setting.key] || setting.default || ''
      return acc
    },
    {}
  )

  return modeSettingValues
}

/** Gets a zoned time object for 03:30 am for the specified date and timezone. */
export function getServiceStart(
  date: string | number | Date,
  timeZone: string
): Date {
  return toDate(`${date} ${SERVICE_BREAK}`, { timeZone })
}

/**
 * Helper function to add/modify parameters from the URL bar
 * while preserving the other ones.
 */
export function combineQueryParams(
  addedParams?: Record<string, unknown>
): string {
  const search = {
    ...getUrlParams(),
    ...addedParams
  }
  return qs.stringify(search, { arrayFormat: 'repeat' })
}
