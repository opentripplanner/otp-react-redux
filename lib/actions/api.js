/* globals fetch */

import { createAction } from 'redux-actions'
import qs from 'qs'
import moment from 'moment'

import { queryIsValid } from '../util/state'
import queryParams from '../util/query-params'
import { updateOtpUrlParams } from '../util/query'
import { hasCar } from '../util/itinerary'

if (typeof (fetch) === 'undefined') require('isomorphic-fetch')

// Generic API actions

export const nonRealtimeRoutingResponse = createAction('NON_REALTIME_ROUTING_RESPONSE')
export const routingRequest = createAction('ROUTING_REQUEST')
export const routingResponse = createAction('ROUTING_RESPONSE')
export const routingError = createAction('ROUTING_ERROR')

let lastSearchId = 0

export function routingQuery () {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const routingType = otpState.currentQuery.routingType
    const searchId = ++lastSearchId

    if (!queryIsValid(otpState)) return
    dispatch(routingRequest({ routingType, searchId }))

    // fetch a realtime route
    fetch(constructRoutingQuery(otpState))
      .then(getJsonAndCheckResponse)
      .then(json => {
        dispatch(routingResponse({ response: json, searchId }))
      })
      .catch(error => {
        dispatch(routingError({ error, searchId }))
      })

    // also fetch a non-realtime route
    fetch(constructRoutingQuery(otpState, true))
      .then(getJsonAndCheckResponse)
      .then(json => {
        dispatch(nonRealtimeRoutingResponse({ response: json, searchId }))
      })
      .catch(error => {
        console.error(error)
        // do nothing
      })
  }
}

function getJsonAndCheckResponse (res) {
  if (res.status >= 400) {
    const error = new Error('Received error from server')
    error.response = res
    throw error
  }
  return res.json()
}

function constructRoutingQuery (otpState, ignoreRealtimeUpdates) {
  const { config, currentQuery } = otpState
  const routingType = currentQuery.routingType
  const isItinerary = routingType === 'ITINERARY'
  const api = isItinerary ? config.api : config.profileApi
  const planEndpoint = `${api.host}${api.port
    ? ':' + api.port
    : ''}${api.path}/plan`

  let params = {}

  // Start with the universe of OTP parameters defined in query-params.js:
  queryParams
    .filter(qp => {
      // A given parameter is included in the request if all of the following:
      // 1. Must apply to the active routing type (ITINERARY or PROFILE)
      // 2. Must be included in the current user-defined query
      // 3. Must pass the parameter's applicability test, if one is specified
      return qp.routingTypes.indexOf(routingType) !== -1 &&
        qp.name in currentQuery &&
        (typeof qp.applicable !== 'function' || qp.applicable(currentQuery))
    })
    .forEach(qp => {
      // Translate the applicable parameters according to their rewrite
      // functions (if provided)
      const rewriteFunction = isItinerary
        ? qp.itineraryRewrite
        : qp.profileRewrite
      params = Object.assign(
        params,
        rewriteFunction
          ? rewriteFunction(currentQuery[qp.name])
          : { [qp.name]: currentQuery[qp.name] }
      )
    })

  // Additional processing specific to ITINERARY mode
  if (isItinerary) {
    // override ignoreRealtimeUpdates if provided
    if (typeof ignoreRealtimeUpdates === 'boolean') {
      params.ignoreRealtimeUpdates = ignoreRealtimeUpdates
    }

    // check date/time validity; ignore both if either is invalid
    const dateValid = moment(params.date, 'YYYY-MM-DD').isValid()
    const timeValid = moment(params.time, 'H:mm').isValid()

    if (!dateValid || !timeValid) {
      delete params.time
      delete params.date
    }

  // Additional processing specific to PROFILE mode
  } else {
    // check start and end time validity; ignore both if either is invalid
    const startTimeValid = moment(params.startTime, 'H:mm').isValid()
    const endTimeValid = moment(params.endTime, 'H:mm').isValid()

    if (!startTimeValid || !endTimeValid) {
      delete params.startTimeValid
      delete params.endTimeValid
    }
  }

  // hack to add walking to driving/TNC trips
  if (hasCar(params.mode)) {
    params.mode += ',WALK'
  }

  // TODO: check that valid from/to locations are provided

  // FIXME: This is only performed when ignoring realtimeupdates currently, just
  // to ensure it is not repeated twice.
  if (ignoreRealtimeUpdates) updateOtpUrlParams(params)

  return `${planEndpoint}?${qs.stringify(params)}`
}

// bike rental station query

export const receivedBikeRentalError = createAction('BIKE_RENTAL_ERROR')
export const receivedBikeRentalResponse = createAction('BIKE_RENTAL_RESPONSE')
export const requestBikeRentalResponse = createAction('BIKE_RENTAL_REQUEST')

export function bikeRentalQuery () {
  return async function (dispatch, getState) {
    const otpState = getState().otp

    dispatch(requestBikeRentalResponse())
    const api = otpState.config.api
    const planEndpoint = `${api.host}${api.port ? ':' + api.port : ''}${api.path}/bike_rental`
    const url = planEndpoint // TODO: add bbox

    let json
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      json = await response.json()
    } catch (err) {
      return dispatch(receivedBikeRentalError(err))
    }

    dispatch(receivedBikeRentalResponse(json))
  }
}

// Single stop lookup query

export const findStopResponse = createAction('FIND_STOP_RESPONSE')
export const findStopError = createAction('FIND_STOP_ERROR')

export function findStop (params) {
  return createQueryAction(`index/stops/${params.stopId}`,
    findStopResponse, findStopError, null, (payload, dispatch) => {
      dispatch(findRoutesAtStop(params.stopId))
      dispatch(findStopTimesForStop({stopId: params.stopId}))
    }
  )
}

// Single trip lookup query

export const findTripResponse = createAction('FIND_TRIP_RESPONSE')
export const findTripError = createAction('FIND_TRIP_ERROR')

export function findTrip (params) {
  return createQueryAction(`index/trips/${params.tripId}`,
    findTripResponse, findTripError, null, (payload, dispatch) => {
      dispatch(findStopsForTrip({tripId: params.tripId}))
      dispatch(findStopTimesForTrip({tripId: params.tripId}))
      dispatch(findGeometryForTrip({tripId: params.tripId}))
    }
  )
}

// Stops for trip query

export const findStopsForTripResponse = createAction('FIND_STOPS_FOR_TRIP_RESPONSE')
export const findStopsForTripError = createAction('FIND_STOPS_FOR_TRIP_ERROR')

export function findStopsForTrip (params) {
  return createQueryAction(`index/trips/${params.tripId}/stops`,
    findStopsForTripResponse, findStopsForTripError,
    (payload) => {
      return {
        tripId: params.tripId,
        stops: payload
      }
    }
  )
}

// Stop times for trip query

export const findStopTimesForTripResponse = createAction('FIND_STOP_TIMES_FOR_TRIP_RESPONSE')
export const findStopTimesForTripError = createAction('FIND_STOP_TIMES_FOR_TRIP_ERROR')

export function findStopTimesForTrip (params) {
  return createQueryAction(`index/trips/${params.tripId}/stoptimes`,
    findStopTimesForTripResponse, findStopTimesForTripError,
    (payload) => {
      return {
        tripId: params.tripId,
        stopTimes: payload
      }
    }
  )
}

// Geometry for trip query

export const findGeometryForTripResponse = createAction('FIND_GEOMETRY_FOR_TRIP_RESPONSE')
export const findGeometryForTripError = createAction('FIND_GEOMETRY_FOR_TRIP_ERROR')

export function findGeometryForTrip (params) {
  return createQueryAction(`index/trips/${params.tripId}/geometry`,
    findGeometryForTripResponse, findGeometryForTripError,
    (payload) => {
      return {
        tripId: params.tripId,
        geometry: payload
      }
    }
  )
}

// Stop times for stop query

export const findStopTimesForStopResponse = createAction('FIND_STOP_TIMES_FOR_STOP_RESPONSE')
export const findStopTimesForStopError = createAction('FIND_STOP_TIMES_FOR_STOP_ERROR')

export function findStopTimesForStop (params) {
  return createQueryAction(`index/stops/${params.stopId}/stoptimes?timeRange=14400`,
    findStopTimesForStopResponse, findStopTimesForStopError,
    (payload) => {
      return {
        stopId: params.stopId,
        stopTimes: payload
      }
    }
  )
}

// Routes lookup query

export const findRoutesResponse = createAction('FIND_ROUTES_RESPONSE')
export const findRoutesError = createAction('FIND_ROUTES_ERROR')

export function findRoutes (params) {
  return createQueryAction('index/routes',
    findRoutesResponse, findRoutesError,
    (payload) => {
      const routes = {}
      payload.forEach(rte => { routes[rte.id] = rte })
      return routes
    }
  )
}

// Single Route lookup query

export const findRouteResponse = createAction('FIND_ROUTE_RESPONSE')
export const findRouteError = createAction('FIND_ROUTE_ERROR')

export function findRoute (params) {
  return createQueryAction(`index/routes/${params.routeId}`,
    findRouteResponse, findRouteError, null,
    (payload, dispatch) => {
      // load patterns
      dispatch(findPatternsForRoute({ routeId: params.routeId }))
    }
  )
}

// Patterns for Route lookup query

export const findPatternsForRouteResponse = createAction('FIND_PATTERNS_FOR_ROUTE_RESPONSE')
export const findPatternsForRouteError = createAction('FIND_PATTERNS_FOR_ROUTE_ERROR')

export function findPatternsForRoute (params) {
  return createQueryAction(`index/routes/${params.routeId}/patterns`,
    findPatternsForRouteResponse, findPatternsForRouteError,
    (payload) => {
      // convert pattern array to ID-mapped object
      const patterns = {}
      payload.forEach(ptn => { patterns[ptn.id] = ptn })

      return {
        routeId: params.routeId,
        patterns
      }
    },
    (payload, dispatch) => {
      // load geometry for each pattern
      payload.forEach(ptn => {
        dispatch(findGeometryForPattern({
          routeId: params.routeId,
          patternId: ptn.id
        }))
      })
    }
  )
}

// Geometry for Pattern lookup query

export const findGeometryForPatternResponse = createAction('FIND_GEOMETRY_FOR_PATTERN_RESPONSE')
export const findGeometryForPatternError = createAction('FIND_GEOMETRY_FOR_PATTERN_ERROR')

export function findGeometryForPattern (params) {
  return createQueryAction(`index/patterns/${params.patternId}/geometry`,
    findGeometryForPatternResponse, findGeometryForPatternError,
    (payload) => {
      return {
        routeId: params.routeId,
        patternId: params.patternId,
        geometry: payload
      }
    }
  )
}

// TNC ETA estimate lookup query

export const transportationNetworkCompanyEtaResponse = createAction('TNC_ETA_RESPONSE')
export const transportationNetworkCompanyEtaError = createAction('TNC_ETA_ERROR')

export function getTransportationNetworkCompanyEtaEstimate (params) {
  const {companies, from} = params
  return createQueryAction(
    `transportation_network_company/eta_estimate?${qs.stringify({
      companies,
      from
    })}`, // endpoint
    transportationNetworkCompanyEtaResponse, // responseAction
    transportationNetworkCompanyEtaError, // errorAction
    (payload) => { // rewritePayload
      return {
        from,
        estimates: payload.estimates
      }
    }
  )
}

// TNC ride estimate lookup query

export const transportationNetworkCompanyRideResponse = createAction('TNC_RIDE_RESPONSE')
export const transportationNetworkCompanyRideError = createAction('TNC_RIDE_ERROR')

export function getTransportationNetworkCompanyRideEstimate (params) {
  const {company, from, rideType, to} = params
  return createQueryAction(
    `transportation_network_company/ride_estimate?${qs.stringify({
      company,
      from,
      rideType,
      to
    })}`, // endpoint
    transportationNetworkCompanyRideResponse, // responseAction
    transportationNetworkCompanyRideError, // errorAction
    (payload) => { // rewritePayload
      return {
        company,
        from,
        rideEstimate: payload.rideEstimate,
        to
      }
    }
  )
}

// generic helper for constructing API queries

function createQueryAction (endpoint, responseAction, errorAction, rewritePayload, postprocess) {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const api = otpState.config.api
    const url = `${api.host}${api.port ? ':' + api.port : ''}${api.path}/${endpoint}`
    let payload
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      payload = await response.json()
    } catch (err) {
      return dispatch(errorAction(err))
    }

    if (typeof rewritePayload === 'function') {
      dispatch(responseAction(rewritePayload(payload)))
    } else {
      dispatch(responseAction(payload))
    }

    if (typeof postprocess === 'function') {
      postprocess(payload, dispatch, getState)
    }
  }
}

// Nearby Stops Query

const receivedNearbyStopsResponse = createAction('NEARBY_STOPS_RESPONSE')
const receivedNearbyStopsError = createAction('NEARBY_STOPS_ERROR')

export function findNearbyStops (params) {
  return createQueryAction(`index/stops?${qs.stringify({radius: 1000, ...params})}`,
    receivedNearbyStopsResponse, receivedNearbyStopsError,
    stops => ({stops}),
    // retrieve routes for each stop
    (stops, dispatch, getState) => stops
      .forEach(stop => dispatch(findRoutesAtStop(stop.id)))
  )
}

// Routes at Stop query

const receivedRoutesAtStopResponse = createAction('ROUTES_AT_STOP_RESPONSE')
const receivedRoutesAtStopError = createAction('ROUTES_AT_STOP_ERROR')

export function findRoutesAtStop (stopId) {
  return createQueryAction(`index/stops/${stopId}/routes`,
    receivedRoutesAtStopResponse, receivedRoutesAtStopError,
    routes => ({stopId, routes})
  )
}

// Stops within Bounding Box Query

const receivedStopsWithinBBoxResponse = createAction('STOPS_WITHIN_BBOX_RESPONSE')
const receivedStopsWithinBBoxError = createAction('STOPS_WITHIN_BBOX_ERROR')

export function findStopsWithinBBox (params) {
  return createQueryAction(`index/stops?${qs.stringify(params)}`,
    receivedStopsWithinBBoxResponse, receivedStopsWithinBBoxError,
    stops => ({stops})
  )
}

export const clearStops = createAction('CLEAR_STOPS_OVERLAY')
