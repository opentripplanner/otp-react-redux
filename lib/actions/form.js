import debounce from 'lodash.debounce'
import qs from 'qs'
import moment from 'moment'
import { createAction } from 'redux-actions'

import { routingQuery } from './api'
import { coordsToString, stringToCoords } from '../util/map'
import { queryIsValid } from '../util/state'
import { getCurrentTime, getCurrentDate } from '../util/time'

export const settingQueryParam = createAction('SET_QUERY_PARAM')
export const changingForm = createAction('FORM_CHANGED')

/**
 * Action to update any specified query parameter. Replaces previous series of
 * parameter-specific actions.
 */
export function setQueryParam (payload) {
  return function (dispatch, getState) {
    dispatch(settingQueryParam(payload))
    dispatch(formChanged())
  }
}

export function parseUrlQueryString (queryString) {
  return function (dispatch, getState) {
    // Trim the leading question mark
    const params = qs.parse(queryString.substring(1))

    // Store params in the current browser history state
    // FIXME: perhaps store this in redux store and never refer to window.history.state
    window.history.replaceState(params, '', queryString)

    // Filter out the OTP (i.e. non-UI) params and set the initial query
    const planParams = {}
    Object.keys(params).forEach(key => {
      if (!key.startsWith('ui_')) planParams[key] = params[key]
    })
    // Convert strings to numbers/objects and dispatch
    dispatch(setQueryParam(planParamsToQuery(planParams)))
  }
}

function planParamsToQuery (params) {
  const query = {}
  for (var key in params) {
    switch (key) {
      case 'fromPlace':
        const from = stringToCoords(params.fromPlace)
        query.from = from.length ? {
          name: coordsToString(from) || null,
          lat: from[0] || null,
          lon: from[1] || null
        } : null
        break
      case 'toPlace':
        const to = stringToCoords(params.toPlace)
        query.to = to.length ? {
          name: coordsToString(to) || null,
          lat: to[0] || null,
          lon: to[1] || null
        } : null
        break
      case 'arriveBy':
        query.departArrive = params.arriveBy === 'true'
          ? 'ARRIVE'
          : params.arriveBy === 'false'
          ? 'DEPART'
          : 'NOW'
        break
      case 'date':
        query.date = params.date || getCurrentDate()
        break
      case 'time':
        query.time = params.time || getCurrentTime()
        break
      default:
        query[key] = params[key]
    }
  }
  return query
}

let debouncedPlanTrip  // store as variable here, so it can be reused.
let lastDebouncePlanTimeMs

export function formChanged () {
  return function (dispatch, getState) {
    const otpState = getState().otp
    // If departArrive is set to 'NOW', update the query time to current
    if (otpState.currentQuery && otpState.currentQuery.departArrive === 'NOW') {
      dispatch(settingQueryParam({ time: moment().format('HH:mm') }))
    }

    dispatch(changingForm())
    const {autoPlan, debouncePlanTimeMs} = otpState.config
    // check if a trip plan should be made
    if (autoPlan && queryIsValid(otpState)) {
      // trip plan should be made

      // check if debouncing function needs to be (re)created
      if (!debouncedPlanTrip || lastDebouncePlanTimeMs !== debouncePlanTimeMs) {
        debouncedPlanTrip = debounce(() => dispatch(routingQuery()), debouncePlanTimeMs)
        lastDebouncePlanTimeMs = debouncePlanTimeMs
      }
      debouncedPlanTrip()
    }
  }
}
