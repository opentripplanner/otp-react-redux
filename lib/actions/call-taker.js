import { createAction } from 'redux-actions'
import { getUrlParams } from '@opentripplanner/core-utils/lib/query'
import { serialize } from 'object-to-formdata'
import qs from 'qs'

import { getISOLikeTimestamp } from '../util/state'
import { isModuleEnabled, Modules } from '../util/config'
import { searchToQuery, sessionIsInvalid } from '../util/call-taker'

import { resetForm } from './form'

if (typeof fetch === 'undefined') require('isomorphic-fetch')

/// PRIVATE ACTIONS

const endingCall = createAction('END_CALL')
const receivedCalls = createAction('RECEIVED_CALLS')
const receivedQueries = createAction('RECEIVED_QUERIES')
const requestingCalls = createAction('REQUESTING_CALLS')
const requestingQueries = createAction('REQUESTING_QUERIES')
const storeSession = createAction('STORE_SESSION')
const toggleFieldTrips = createAction('TOGGLE_FIELD_TRIPS')

/// PUBLIC ACTIONS

export const beginCall = createAction('BEGIN_CALL')
export const toggleCallHistory = createAction('TOGGLE_CALL_HISTORY')
export const toggleMailables = createAction('TOGGLE_MAILABLES')

/**
 * Fully reset form and toggle call history (and close field trips if open).
 */
export function resetAndToggleCallHistory() {
  return function (dispatch, getState) {
    dispatch(resetForm(true))
    dispatch(toggleCallHistory())
    if (getState().callTaker.fieldTrip.visible) dispatch(toggleFieldTrips())
  }
}

/**
 * Start a new call (and show the call history window) if
 * - the call module is enabled, and
 * - a call is not in progress, and
 * - the field trip window is not visible.
 */
export function beginCallIfNeeded() {
  return function (dispatch, getState) {
    const state = getState()
    const { activeCall, fieldTrip } = state.callTaker
    const callTakerEnabled = isModuleEnabled(state, Modules.CALL_TAKER)
    if (callTakerEnabled && !activeCall && !fieldTrip.visible) {
      dispatch(beginCall())
    }
  }
}

/**
 * Handle initializing a new Trinet session by redirecting to Trinet auth and
 * returning once authenticated successfully.
 */
function newSession(datastoreUrl, verifyLoginUrl, redirect) {
  fetch(datastoreUrl + '/auth/newSession')
    .then((res) => res.json())
    .then((data) => {
      const { sessionId: session } = data
      const windowUrl = `${verifyLoginUrl}?${qs.stringify({
        redirect,
        session
      })}`
      // Redirect to login url.
      window.location = windowUrl
    })
    .catch((error) => {
      console.error('newSession error', error)
    })
}

/**
 * Check that a particular session ID is valid and store resulting session
 * data.
 */
function checkSession(datastoreUrl, sessionId, intl) {
  return function (dispatch, getState) {
    fetch(datastoreUrl + `/auth/checkSession?sessionId=${sessionId}`)
      .then((res) => res.json())
      .then((session) => dispatch(storeSession({ session })))
      .catch((err) => {
        dispatch(storeSession({ session: null }))
        alert(
          intl.formatMessage(
            { id: 'actions.callTaker.checkSessionError' },
            { err: JSON.stringify(err) }
          )
        )
      })
  }
}

/**
 * Initialize the Call Taker and Field Trip modules by checking the session
 * query param against sessions in the datastore backend or initializing a new
 * session via Trinet.
 */
export function initializeModules(intl) {
  return function (dispatch, getState) {
    const { datastoreUrl, trinetReDirect } = getState().otp.config
    // Initialize session if datastore enabled.
    if (datastoreUrl && trinetReDirect) {
      // TODO: Generalize for non-TriNet instances.
      const sessionId = getUrlParams().sessionId
      if (sessionId) {
        // Initialize the session if found in URL query params.
        dispatch(checkSession(datastoreUrl, sessionId, intl))
      } else {
        // No sessionId was passed in, so we must request one from server.
        newSession(datastoreUrl, trinetReDirect, window.location.origin)
      }
    }
  }
}

/**
 * Fetch latest calls for a particular session.
 */
export function fetchCalls(intl) {
  return async function (dispatch, getState) {
    dispatch(requestingCalls())
    const { callTaker, otp } = getState()
    if (sessionIsInvalid(callTaker.session)) return
    const { datastoreUrl } = otp.config
    const { sessionId } = callTaker.session
    const limit = 1000
    try {
      const fetchResult = await fetch(
        `${datastoreUrl}/calltaker/call?${qs.stringify({ limit, sessionId })}`
      )
      const calls = await fetchResult.json()
      dispatch(receivedCalls({ calls }))
    } catch (err) {
      alert(
        intl.formatMessage(
          { id: 'actions.callTaker.fetchCallsError' },
          { err: JSON.stringify(err) }
        )
      )
    }
  }
}

/**
 * Store the trip queries made over the course of a call (to be called when the
 * call terminates).
 */
function saveQueriesForCall(call, intl) {
  return function (dispatch, getState) {
    const { callTaker, otp } = getState()
    const { datastoreUrl } = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    if (!call) {
      alert(
        intl.formatMessage({ id: 'actions.callTaker.couldNotFindCallError' })
      )
      return
    }
    return Promise.all(
      call.searches.map((searchId) => {
        const search = otp.searches[searchId]
        const query = searchToQuery(search, call, otp.config)
        const { sessionId } = callTaker.session
        const queryData = serialize({ query, sessionId })
        return fetch(
          `${datastoreUrl}/calltaker/callQuery?sessionId=${sessionId}`,
          { body: queryData, method: 'POST' }
        )
          .then((res) => res.json())
          .catch((err) => {
            alert(
              intl.formatMessage(
                { id: 'actions.callTaker.callQuerySaveError' },
                { err: JSON.stringify(err) }
              )
            )
          })
      })
    )
  }
}

/**
 * End the active call and store the queries made during the call.
 */
export function endCall(intl) {
  return async function (dispatch, getState) {
    const { callTaker, otp } = getState()
    const { activeCall, session } = callTaker
    const { sessionId } = session
    if (sessionIsInvalid(session)) return
    // Make POST request to store new call.
    const callData = serialize({
      call: {
        endTime: getISOLikeTimestamp(otp.config.homeTimezone),
        startTime: activeCall.startTime
      },
      sessionId
    })
    try {
      const fetchResult = await fetch(
        `${otp.config.datastoreUrl}/calltaker/call`,
        { body: callData, method: 'POST' }
      )
      const id = await fetchResult.json()

      // Inject call ID into active call and save queries.
      await dispatch(saveQueriesForCall({ ...activeCall, id }, intl))
      // Wait until query was saved before re-fetching queries for this call.
      await dispatch(fetchCalls(intl))
    } catch (err) {
      console.error(err)
      alert(
        intl.formatMessage(
          { id: 'actions.callTaker.callSaveError' },
          { err: JSON.stringify(err) }
        )
      )
    }
    // Clear itineraries shown when ending call.
    dispatch(resetForm(true))
    dispatch(endingCall())
  }
}

/**
 * Fetch the trip queries that were made during a particular call.
 */
export function fetchQueries(callId, intl) {
  return function (dispatch, getState) {
    dispatch(requestingQueries())
    const { callTaker, otp } = getState()
    const { datastoreUrl } = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const { sessionId } = callTaker.session
    fetch(
      `${datastoreUrl}/calltaker/callQuery?sessionId=${sessionId}&call.id=${callId}`
    )
      .then((res) => res.json())
      .then((queries) => {
        dispatch(receivedQueries({ callId, queries }))
      })
      .catch((err) => {
        alert(
          intl.formatMessage(
            { id: 'actions.callTaker.queryFetchError' },
            { err: JSON.stringify(err) }
          )
        )
      })
  }
}
