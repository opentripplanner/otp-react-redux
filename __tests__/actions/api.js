/* globals describe, expect, it, jest */

import nock from 'nock'

import {routingQuery} from '../../lib/actions/api'

describe('actions > api', () => {
  describe('> routingQuery', () => {
    const defaultState = {
      otp: {
        config: {
          api: {
            host: 'http://mock-host.com',
            path: '/api',
            port: 80
          }
        },
        currentQuery: {
          from: { lat: 12, lon: 34 },
          mode: 'WALK,TRANSIT',
          routingType: 'ITINERARY',
          to: { lat: 34, lon: 12 }
        },
        searches: []
      }
    }

    it('should make a query to OTP', async () => {
      const routingQueryAction = routingQuery()

      nock('http://mock-host.com')
        .get(/api\/plan/)
        .reply(200, (uri, requestBody) => {
          expect(uri).toMatchSnapshot('OTP Query Path')
          return { fake: 'response' }
        })

      const mockDispatch = jest.fn()
      await routingQueryAction(mockDispatch, () => {
        return defaultState
      })

      expect(mockDispatch.mock.calls).toMatchSnapshot()
    })

    it('should gracefully handle bad response', async () => {
      const routingQueryAction = routingQuery()

      nock('http://mock-host.com')
        .get(/api\/plan/)
        .reply(500, {
          fake: 'response'
        })

      const mockDispatch = jest.fn()
      await routingQueryAction(mockDispatch, () => {
        return defaultState
      })

      expect(mockDispatch.mock.calls).toMatchSnapshot()
    })
  })
})
