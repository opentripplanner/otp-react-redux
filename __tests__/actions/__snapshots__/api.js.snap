// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTP Query Path 1`] = `"/api/plan?fromPlace=12%2C34&toPlace=34%2C12&mode=WALK%2CTRANSIT"`;

exports[`actions > api > routingQuery should gracefully handle bad response 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "routingType": "ITINERARY",
        "searchId": 2,
      },
      "type": "ROUTING_REQUEST",
    },
  ],
]
`;

exports[`actions > api > routingQuery should make a query to OTP 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "routingType": "ITINERARY",
        "searchId": 1,
      },
      "type": "ROUTING_REQUEST",
    },
  ],
]
`;
